### ISSUE TITLE
regression-report fails because required cache is not saved from issue_comment workflow

### ISSUE THREAD

**Issue by gianmarcoschifone** [checkstyle/checkstyle#20538]
## Problem

Detected at https://github.com/checkstyle/checkstyle/pull/20477.

The `regression-report` workflow fails before it can parse the trigger comment or generate a report.

The failure happens in [.github/workflows/regression-report.yml](https://github.com/checkstyle/checkstyle/blob/master/.github/workflows/regression-report.yml), where the workflow uses GitHub Actions cache as required storage between jobs.

## Evidence

The workflow is triggered by `issue_comment`:

```yaml
on:
  issue_comment:
    types: [ created ]
```

It creates a new cache key for each comment:

```yaml
CACHE_KEY: "regression-report-${{ github.sha }}-${{ github.event.comment.id }}"
```

The `checkout_and_cache` job saves the workspace with that key:

```yaml
- name: Save cache
  uses: actions/cache/save@v6
  with:
    path: .
    key: ${{ env.CACHE_KEY }}
```

In failing runs, GitHub rejects the cache write:

```text
Failed to save: Unable to reserve cache with key regression-report-01f75536e1b3d283c16d9425bf3163807592d634-4850570559.
More details: cache write denied: token has no writable scopes
```

The next job restores the same key with `fail-on-cache-miss: true`:

```yaml
- name: Restore cache
  uses: actions/cache/restore@v6
  with:
    path: .
    key: ${{ env.CACHE_KEY }}
    fail-on-cache-miss: true
```

Since the cache was not created, restore fails:

```text
Failed to restore cache entry. Exiting as fail-on-cache-miss is set.
Input key: regression-report-01f75536e1b3d283c16d9425bf3163807592d634-4850570559
```

The workflow already has `actions: write`, so this is not fixed by adding workflow permissions:

```yaml
permissions:
  contents: read
  pull-requests: write
  actions: write
```

## GitHub Cache Policy Context

GitHub added official cache security documentation on `2026-06-30 00:27:21 UTC`:

- https://github.com/github/docs/commit/215d2a0c1ca8ee2d9ca9ebaab266364d42eb09ec

The current documentation says that some low-trust workflow triggers, including `issue_comment` on the default branch, have read-only cache access:

- https://docs.github.com/en/actions/reference/workflows-and-actions/dependency-caching#cache-access-for-low-trust-workflow-triggers

This matches the observed failure: the workflow can try to save a cache, but GitHub denies the write with:

```text
cache write denied: token has no writable scopes
```

Verified Checkstyle window:

- Cache handoff still worked in run `28374335079`, created at `2026-06-29 13:08:20 UTC`: https://github.com/checkstyle/checkstyle/actions/runs/28374335079
- Cache handoff was already failing in run `28430277714`, created at `2026-06-30 08:17:00 UTC`: https://github.com/checkstyle/checkstyle/actions/runs/28430277714

So, for Checkstyle, this behavior changed between `2026-06-29 13:08:20 UTC` and `2026-06-30 08:17:00 UTC`.

## Cause

`regression-report.yml` uses dependency cache as mandatory inter-job storage.

That is no longer reliable for this workflow because:

- the trigger is `issue_comment`;
- each report request uses a unique cache key based on `github.event.comment.id`;
- the cache cannot already exist;
- GitHub now denies cache writes for this low-trust trigger.

The save failure is only a warning, so `checkout_and_cache` appears successful. Later jobs fail because they require the missing cache.

## Proposed Solution

Replace the required cache handoffs with workflow artifacts.

GitHub documents `upload-artifact` and `download-artifact` as the mechanism for passing data between jobs in the same workflow. Dependency cache should remain for optional dependency reuse, such as Maven cache, not for required per-run state.

Suggested artifact names:

```yaml
env:
  WORKSPACE_ARTIFACT: regression-workspace-${{ github.run_id }}-${{ github.event.comment.id }}
  GENERATED_CONFIGS_ARTIFACT: regression-generated-configs-${{ github.run_id }}-${{ github.event.comment.id }}
  CONFIG_FILES_ARTIFACT: regression-config-files-${{ github.run_id }}-${{ github.event.comment.id }}
```

Replace workspace cache save with:

```yaml
- name: Upload workspace
  uses: actions/upload-artifact@v7
  with:
    name: ${{ env.WORKSPACE_ARTIFACT }}
    path: .
    retention-days: 1
    if-no-files-found: error
```

Replace workspace cache restore with:

```yaml
- name: Download workspace
  uses: actions/download-artifact@v8
  with:
    name: ${{ env.WORKSPACE_ARTIFACT }}
    path: .
```

The same conversion should be done for generated configs and config files.

## Links

- Failing save cache job: https://github.com/checkstyle/checkstyle/actions/runs/28495906196/job/84462016818
- Failing restore cache job: https://github.com/checkstyle/checkstyle/actions/runs/28495906196/job/84462070231
- First verified Checkstyle cache failure: https://github.com/checkstyle/checkstyle/actions/runs/28430277714
- Last verified Checkstyle run where the workspace cache restored successfully: https://github.com/checkstyle/checkstyle/actions/runs/28374335079
- GitHub cache security docs commit: https://github.com/github/docs/commit/215d2a0c1ca8ee2d9ca9ebaab266364d42eb09ec
- GitHub dependency caching reference: https://docs.github.com/en/actions/reference/workflows-and-actions/dependency-caching
- GitHub artifact handoff documentation: https://docs.github.com/en/actions/tutorials/store-and-share-data#passing-data-between-jobs-in-a-workflow
- `actions/upload-artifact` documentation: https://github.com/actions/upload-artifact
- `actions/download-artifact` documentation: https://github.com/actions/download-artifact

**Comment by romani** [comment]
even after downgrade still same failure https://github.com/checkstyle/checkstyle/actions/runs/28671873198

Something is broken/changed in Github to cause this.

**Comment by romani** [comment]
gianmarco
https://docs.github.com/en/actions/reference/workflows-and-actions/dependency-caching#cache-access-for-low-trust-workflow-they changed 4 days ago
https://github.com/github/docs/commit/215d2a0c1ca8ee2d9ca9ebaab266364d42eb09ec

**Comment by stoyanK7** [comment]
Testing `regression-report.yml` on my own fork is a bit more complicated. I think I will have a fix for the Workflow job by tomorrow.

**Comment by stoyanK7** [comment]
PR
- #20565 

is only for `regression-report.yml`. We still have `diff-report.yml` and `linkcheck-on-comment.yml` to verify.

I have not confirmed whether these 2 are broken, but I believe they might not be since they use `actions/cache@v6` only once for the Maven cache. We still have to test in some PR and confirm, though. Before closing this issue.

**Comment by romani** [comment]
Testing:

> linkcheck-on-comment.yml

https://github.com/checkstyle/checkstyle/pull/20570#issuecomment-4884193459

And it works 
https://github.com/checkstyle/checkstyle/pull/20570#issuecomment-4884202456

**Comment by romani** [comment]
>diff-report.yml

Tested at 
https://github.com/checkstyle/checkstyle/pull/20582#issuecomment-4886863380 works well 

Issue is resolved
